compilers: Pass -no-undefined to GNU ld by default

nirbheek opened the pull request:
Without passing -no-undefined, the linker does not warn about missing symbols at link-time (due to a misconfigured build system, missing dependencies, etc). I found that a lot of modules/libraries in GStreamer were being built incorrectly under Meson and were failing at use-time because of this.

This is the default for all other linkers, and most projects already set this by default. The only reason to not have this is obscure platforms like BeOS and HPPA that use this as a feature. For those, the required symbols should be explicitly set as undefined with the --undefined linker option.

We have to use `-Wl` because we use `gcc` to invoke the linker.

jpakkane commented:
Having this unconditionally on is bad, because address sanitizer requires that it is _not_ used and we want to provide a good out-of-the-box experience with that (the support is not in trunk yet but there is an outstanding mr).

nirbheek commented:
Doesn't that just mean that you need to pass `-lasan` during linking when using the address sanitizer? IIRC, it requires that you don't pass `-z,defs` (because that disallows undefined symbols in object files as well).

jpakkane commented:
Q: When I link my shared library with -fsanitize=address, it fails due to some undefined ASan symbols (e.g. asan_init_v4)?

    A: Most probably you link with -Wl,-z,defs or -Wl,--no-undefined. These flags don't work with ASan.

[Source](https://github.com/google/sanitizers/wiki/AddressSanitizer).

tp-m commented:
I think it is still a good default and desirable in the case where address sanitizer is not used then, so perhaps that needs to be disabled somehow by the address-sanitizer mod then, I don't see why one would decide on the defaults based on some special-case feature?

jpakkane commented:
Something like that would make sense. Maybe have a compiler option like `no-undefined` with values of `auto` and `off`. The former would enable it unless there is a different option that conflicts.

Having an explicit off is also useful, there may be other tools or cases where you might want to disable this argument.

nirbheek commented:
I think that page is wrong. The [original llvm page about this](http://clang.llvm.org/docs/AddressSanitizer.html#usage) only mentions `-z,defs`, and from what I can make out, projects out there that have `-no-undefined` as the default either turn it off if ASAN is used, or add `-lasan` if ASAN is enabled.

In fact, I tested it, and passing `-Wl,-no-undefined -fsanitizer=address -lasan` while linking works fine.

nirbheek commented:
I implemented this with three options: `auto`, `force`, `off`. I added `force` in case projects wanted to, for instance, ignore Meson's (future) auto-disabling of `--no-undefined` for `-faddress=sanitize` and wanted to fix it by other means such as adding `-lasan` or explicitly listing all the undefined symbols.

jpakkane commented:
Filed bug on the sanitizers asking them to clarify their docs on this behaviour. The issue is google/sanitizers#660.

jpakkane commented:
And now we have the official word of $deity. `-lasan` is not supported so that's not something we can do.

jpakkane commented:
Testing proved this to be more of a clusterf*ck than expected. Gcc does not have `--shared-libasan` at all but seems to enable it by default. Putting it there explicitly breaks. On Fedora the necessary libraries for Clang do not seem to be available at all so it does not work and I can't test it. :(

nirbheek commented:
This has been superseded by https://github.com/mesonbuild/meson/pull/454, so closing pull request.
